kindlecomicconverter/comic2ebook.py

```diff
@@ -581,6 +593,69 @@ def buildEPUB(path, chapternames, tomenumber, ischunked, cover: image.Cover, len
     buildOPF(path, options.title, filelist, cover)
 
 
+def buildPDF(path, title, cover=None, output_file=None):
+    """
+    Build a PDF file from processed comic images.
+    Images are combined into a single PDF optimized for e-readers.
+    """
+    import io
+    import gc
+    
+    # prepare image list
+    images_path = os.path.join(path, "OEBPS", "Images")
+    image_files = [os.path.join(d, f) for d, _, files in walkLevel(images_path)
+                   for f in files if f.lower().endswith(('.png', '.jpg', '.jpeg', '.gif', '.webp'))]
+    if not image_files:
+        raise UserWarning("No images found for PDF creation.")
+
+    # sort image files to ensure correct order
+    image_files.sort()
+    
+    # open empty PDF
+    doc = pymupdf.open()
+    
+    # Calculate optimal chunksize based on image count and available cores
+    num_cores = cpu_count()
+    num_images = len(image_files)
+    # Optimal chunksize balances overhead vs parallelism
+    optimal_chunksize = max(1, min(10, num_images // (num_cores * 2)))
+    
+    with Pool(processes=num_cores) as pool:
+        # Process images in parallel but maintain order
+        processed_images = pool.map(_process_pdf_image, image_files, chunksize=optimal_chunksize)
+        
+        # Stream images to PDF
+        for i, img in enumerate(processed_images):
+            if not img:
+                continue
+            
+            w, h = img.size
+            page = doc.new_page(width=w, height=h)
+            
+            with io.BytesIO() as buf:
+                if getattr(options, 'forcepng', False):
+                    img.save(buf, format="PNG", optimize=True)
+                else:
+                    img.save(buf, format="JPEG", quality=90, optimize=True)
+                page.insert_image(page.rect, stream=buf.getvalue())
+            
+            # Explicit cleanup for large images
+            del img
+            
+            # Periodic garbage collection for large comic books to prevent running out of memory
+            if (i + 1) % 50 == 0:
+                gc.collect()
```

Comment:
How big files were you using?

Reply:
350 MB for 30 pages. Some of the people I know have comics of > 1 GB. Also if I am not mistaken, when loaded in memory it's larger because of decompression. At the same time, images are processed in parallel so this kan make the memory go up really fast. F.e. my macbook pro has 14 cores but you can buy it with a very small memory footprint. I am keeping in mind that a lot of people still have 8 GB memory